Content.Shared/Buckle/SharedBuckleSystem.Buckle.cs

```diff
@@ -215,144 +216,150 @@ protected void SetBuckledTo(Entity<BuckleComponent> buckle, Entity<StrapComponen
+                _popup.PopupClient(message, userUid);
             }
 
             return false;
         }
 
-        var buckleAttempt = new BuckleAttemptEvent((strapUid, strapComp), (buckleUid, buckleComp), user, popup);
-        RaiseLocalEvent(buckleUid, ref buckleAttempt);
+        var buckleAttempt = new BuckleAttemptEvent(strap, buckle, userUid, popup);
+        RaiseLocalEvent(buckle.Owner, ref buckleAttempt);
         if (buckleAttempt.Cancelled)
             return false;
 
-        var strapAttempt = new StrapAttemptEvent((strapUid, strapComp), (buckleUid, buckleComp), user, popup);
-        RaiseLocalEvent(strapUid, ref strapAttempt);
+        var strapAttempt = new StrapAttemptEvent(strap, buckle, userUid, popup);
+        RaiseLocalEvent(strap.Owner, ref strapAttempt);
         if (strapAttempt.Cancelled)
             return false;
 
         return true;
     }
 
     /// <summary>
-    /// Attempts to buckle an entity to a strap
+    /// Attempts to buckle an entity to a strapUid
     /// </summary>
-    /// <param name="buckle"> Uid of the owner of BuckleComponent </param>
-    /// <param name="user">
+    /// <param name="buckleUid"> Uid of the owner of BuckleComponent </param>
+    /// <param name="userUid">
     /// Uid of a third party entity,
     /// i.e, the uid of someone else you are dragging to a chair.
     /// Can equal buckleUid sometimes
     /// </param>
-    /// <param name="strap"> Uid of the owner of strap component </param>
-    public bool TryBuckle(EntityUid buckle, EntityUid? user, EntityUid strap, BuckleComponent? buckleComp = null, bool popup = true)
+    /// <param name="strapUid"> Uid of the owner of strapUid component </param>
+    public bool TryBuckle(EntityUid buckleUid, EntityUid? userUid, EntityUid strapUid, BuckleComponent? buckleComp = null, bool popup = true)
     {
-        if (!Resolve(buckle, ref buckleComp, false))
+        if (!Resolve(buckleUid, ref buckleComp, false))
+            return false;
+        var buckle = (buckleUid, buckleComp);
+
+        if (!TryComp<StrapComponent>(strapUid, out var strapComp))
             return false;
+        var strap = (strapUid, strapComp);
 
-        if (!CanBuckle(buckle, user, strap, popup, out var strapComp, buckleComp))
+        if (!CanBuckle(buckle, strap, userUid, popup))
             return false;
 
-        Buckle((buckle, buckleComp), (strap, strapComp), user);
+        if (!InvalidateSelfPulling(buckle, strap))
+            return false;
+
+        Buckle(buckle, strap, userUid);
         return true;
     }
 
+    private bool InvalidateSelfPulling(Entity<BuckleComponent> buckle, Entity<StrapComponent> strap)
+    {
+        if (!TryComp<PullableComponent>(strap.Owner, out var pullable))
+            return true; // strap owner is not pullable, self pulling not possible
+
+        if (pullable.Puller != buckle.Owner)
+            return true;
+
+        return _pullingSystem.TryStopPull(strap.Owner, pullable);
+    }
+
```

Comment:
This should be part of canbuckle because it's possible for CanBuckle to pass and trybuckle to fail which should never happen.

Reply:
So you are tellin me that:
1. TryStopPull can't be part of CanBuckle because its changing state.
2. TryStopPull can't be part of Buckle because it can fail.
3. TryStopPull can't be it's own thing because  if CanBuckle  passes TryBuckle also has to pass.

Ok, I will need to refactor TryStopPulll into CanStopPull and StopPull
then I can do CanStopPull in CanBuckle  and StopPull StopBuckle.

Reply:
IDK about buckle but trystoppull shouldn't be in canbuckle and you should be able to call canbuckle then buckle and not have behavior different to trybuckle

Reply:
As Emo said putting any "Try" method into buckle is not optimal:
"Presently, the current code doesn't handle a situation where the TryStopPull call fails to actually stop the pull (since it's not guaranteed) causing the assert to still trigger"
I can't cancel buckling either as it suppose to buckle after positive verification from CanBuckle